### ISSUE TITLE
Valid TypeScript has "SyntaxError: '=' expected."

### ISSUE THREAD

**Issue by abirmingham** [prettier/prettier#2702]
The following code causes a SyntaxError, despite being syntactically/stylistically correct:
```js
const bar = (...varargs:any[]) => {
  console.log(varargs);
};

const foo = (x:string):void => (
  bar(
    x,
    () => {},
    () => {}
  )
);
```

Here is the result:
```js
SyntaxError: '=' expected. (6:6)
  4 | 
  5 | const foo = (x:string):void => (
> 6 |   bar(
    |      ^
  7 |     x,
  8 |     () => {},
  9 |     () => {}
```

[Here is a link](https://prettier.io/playground/#%7B%22content%22%3A%22const%20bar%20%3D%20(...varargs%3Aany%5B%5D)%20%3D%3E%20%7B%5Cn%20%20console.log(varargs)%3B%5Cn%7D%3B%5Cn%5Cnconst%20foo%20%3D%20(x%3Astring)%3Avoid%20%3D%3E%20(%5Cn%20%20bar(%5Cn%20%20%20%20x%2C%5Cn%20%20%20%20()%20%3D%3E%20%7B%7D%2C%5Cn%20%20%20%20()%20%3D%3E%20%7B%7D%5Cn%20%20)%5Cn)%3B%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22typescript%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Afalse%7D%7D) to the issue reproduced in the prettier playground. Note that if the parser is changed to babylon then there is no error shown.

Thanks for any assistance!

EDIT: I'm using prettier@1.6.1 and typescript@2.2.2.

**Comment by vjeux** [comment]
@JamesHenry do you know what's going on here?

**Comment by JamesHenry** [comment]
Funky, seems to be an issue with the TypeScript 2.5 RC. This parses fine in stable TypeScript 2.4, but not in 2.5.1.

I'll report it to the TS team

**Comment by abirmingham** [comment]
@JamesHenry many thanks!

I should have mentioned that I'm using prettier@1.6.1 and typescript@2.2.2.

**Comment by JamesHenry** [comment]
Prettier bundles its required TypeScript version, so your installed version (of 2.2.2) isn't actually relevant to the issue. Everyone who installs a particular version of prettier gets the same underlying TypeScript parser version.

Here is the bug report: https://github.com/Microsoft/TypeScript/issues/18123

**Comment by alexburner** [comment]
@JamesHenry it seems strange Prettier includes its own version of TypeScript, instead of using the version provided by the hosting project. What if there's a significant/breaking change between TS versions?

**Comment by vjeux** [comment]
@alexburner so far all of the parsers we've made didn't introduce breaking changes. It's extremely onerous for them as they break compatibility with all the codebases written in that language, so they usually end up not doing it.

With prettier we're using versions close to master of the next release, so that when they ship a big new major version, we already have had support for all the new syntax additions for a while.

For prettier, we are hitting the internals of those parsers and none of the parsers are maintaining backward compatibility of the internals, so it would be a significant effort to try and be compatible with arbitrary versions of the parsers that happen to be in your project. It also would make our life harder with integrations as to how to find the right version.

**Comment by alexburner** [comment]
Thanks for the quick response!

> For prettier, we are hitting the internals of those parsers and none of the parsers are maintaining backward compatibility of the internals, so it would be a significant effort to try and be compatible with arbitrary versions of the parsers that happen to be in your project.

That sounds like plenty of reason right there, I hadn't realized your work required customizing the internals of those parsers. That would make full version compatibility much more difficult.

**Comment by JamesHenry** [comment]
@azz I'm on mobile while travelling, so hard to follow up, but the underlying TypeScript issue for this was fixed. Not sure if the TS version was bumped when releasing 1.7.0?

**Comment by azz** [comment]
This is not fixed in `lastest` (2.5.2) but is fixed in `insiders` (`2.5.3-insiders.20170909`)
